pkg/controller/registration/registration_controller.go

```diff
@@ -56,16 +56,17 @@ var logger = klog.Log.WithName(controllerName)
 
 // Add creates a new registration Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
 // and Start it when the Manager is Started.
-func Add(mgr manager.Manager, p *tfschema.Provider, smLoader *servicemappingloader.ServiceMappingLoader, dclConfig *dcl.Config, dclConverter *conversion.Converter, regFunc registrationFunc) error {
+func Add(mgr manager.Manager, p *tfschema.Provider, smLoader *servicemappingloader.ServiceMappingLoader, dclConfig *dcl.Config, dclConverter *conversion.Converter, regFunc registrationFunc, stateIntoSpecValue *k8s.StateIntoSpecValue) error {
```

Comment:
Added the `Defaulter` interface and changed `StateIntoSpecValue` into `StateIntoSpecDefaulter`. If we are going to support more defaulters in the future, we probably can turn it into a defaulter map / struct (instead of appending more parameters).

We'll probably improve the generic interface over time when we support more defaulters.

Also we seem to not really do nil checks for pointers passed over function calls into the reconcilers. Are you suggesting that we should do it?

Reply:
Sorry, I should have been more verbose:

By defaulter I meant a function or interface:
```
type Defaulter interface {
  ApplyDefaults(ctx context.Context, obj client.Object) (changed bool, err error)
}
```

Then you do this:
```
type Reconciler struct {
...
  defaulter Defaulter
...
}

func (r *Reconciler) Reconcile(ctx context.Context...
...
  if defaulter != nil {
    _, err := r.defaulter.ApplyDefaults(ctx, obj); err != nil {
      return nil, err
      } 
  }
...
```

If you're going to do that, then you can avoid the nil check with this:

```
type NoopDefaulter struct {
}

var _ Defaulter = &NoopDefaulter{}

func (d *NoopDefaults) ApplyDefaults(ctx context.Context, obj client.Object) (bool, error) {
  return false, nil
}
```

But easier still may be to do this:
```
type Reconciler struct {
...
  defaulters []Defaulter
...
}

func (r *Reconciler) Reconcile(ctx context.Context...
...
  for _, defaulter := range r.defaulters {
    _, err := defaulter.ApplyDefaults(ctx, obj); err != nil {
      return nil, err
      } 
  }
...
```

Reply:
Thank you for the detailed suggestions! I like the idea of using a list of defaulters. I think it's good to make the interface as generic as possible so I kept `ctx context.Context` in the parameters and `changed bool` in the returned values for `ApplyDefaults()`.